File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int toMinute(int hour, int minute) {
    return hour * 60  + minute;
}

int calcFee(vector<int> fees, int minute) {
    int b_time = fees[0];
    int b_fee = fees[1];
    int u_time = fees[2];
    int u_fee = fees[3];
    if (minute <= b_time) return b_fee;
    return b_fee + ((minute - b_time + u_time - 1) / u_time) * u_fee;
}

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    map<string, int> inTime;
    map<string, int> totalTime;
    for (auto record : records) {
        int hour = stoi(record.substr(0, 2));
        int minute = stoi(record.substr(3, 2));
        string number = record.substr(6, 4);
        string status = record.substr(11);
        
        int currentTime = toMinute(hour, minute);        
        int lastInTime = inTime[number];
        
        if (status == "IN") {
            inTime[number] = currentTime;
        }
        if (status == "OUT") {
            totalTime[number] += currentTime - lastInTime;
            inTime.erase(number);
        }             
    }
    
    int endTime = toMinute(23, 59);
    for (auto enter: inTime) {
        totalTime[enter.first] += endTime - enter.second;
    }
    
    for (auto io: totalTime) {
        int time = io.second;
        answer.push_back(calcFee(fees, time));
    }
    
    return answer;
    
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math

def to_minute(timestamp):
    h, m = timestamp.split(":")
    return int(h) * 60 + int(m)
    
    
def solution(fees, records):
    min_time, min_fee, unit_time, unit_fee = fees
    
    def calc_fee(time):
        if time <= min_time:
            return min_fee
        return min_fee + math.ceil((time - min_time) / unit_time) * unit_fee 
    
    in_time = dict()
    total_time = dict()
    
    for record in records:
        timestamp, number, status = record.split()
        current_time = to_minute(timestamp)
        
        if in_time.get(number, -1) < 0:
            in_time[number] = current_time
        else:
            total_time[number] = total_time.get(number, 0) + (current_time - in_time[number])
            in_time.pop(number)
    
    end_time = to_minute("23:59")
    for number in in_time:
        total_time[number] = total_time.get(number, 0) + (end_time - in_time[number])
        
    key_list = sorted(list(total_time.keys()))
    answer = []
    for key in key_list:
        answer.append(calc_fee(total_time[key]))
    return answer